### PR TITLE
Fp 69

### PR DESCRIPTION
**Pull request by Archifer**

### CODE REVIEW

**Review by PiotrKlis — APPROVED** [review]
fine from android side

File: ios/Classes/NativeBridge/Handlers/MobileAuthHandler.swift [hunk 1]
```diff
@@ -2,132 +2,128 @@ import Foundation
...
-
-enum MobileAuthAuthenticatorType: String {
-    case fingerprint = "biometric"
-    case pin = "PIN"
-    case confirmation = ""
+    func handleMobileAuthWithOtp(otp: String, completion: @escaping (Result<Void, FlutterError>) -> Void)
+    func enrollMobileAuthentication(completion: @escaping (Result<Void, FlutterError>) -> Void)
+    func acceptMobileAuthRequest(completion: @escaping (Result<Void, FlutterError>) -> Void)
+    func denyMobileAuthRequest(completion: @escaping (Result<Void, FlutterError>) -> Void)
+    func setMobileAuthCallback(didReceiveConfirmationChallenge confirmation: @escaping (Bool) -> Void)
```
Comment (szymon-onewelcome) [inline ios/Classes/NativeBridge/Handlers/MobileAuthHandler.swift:9]: You don't need `didReceiveConfirmationChallenge`, just  `func setMobileAuthCallback(_ confirmation: @escaping (Bool) -> Void)`. 
BTW
1. why do you need extra "callback" if every method has completion?
2. why bool insted of some result?

File: ios/Classes/NativeBridge/Handlers/MobileAuthHandler.swift [hunk 2]
```diff
@@ -2,132 +2,128 @@ import Foundation
...
-                    return;
-                }
-                
-                completion(true, nil)
-              }
+    func handleMobileAuthWithOtp(otp: String, completion: @escaping (Result<Void, FlutterError>) -> Void) {
+        Logger.log("handleMobileAuthWithOtp", sender: self)
+
+        // Check to prevent breaking iOS SDK; https://onewelcome.atlassian.net/browse/SDKIOS-987
+        guard let _ = SharedUserClient.instance.authenticatedUserProfile else {
```
Comment (szymon-onewelcome) [inline ios/Classes/NativeBridge/Handlers/MobileAuthHandler.swift:23]: if `SharedUserClient.instance.authenticatedUserProfile` is marked as `@discardableResult` you don't need `_ =` or you can check 
```
guard SharedUserClient.instance.authenticatedUserProfile == nil else {
...

File: ios/Classes/NativeBridge/Handlers/MobileAuthHandler.swift [hunk 3]
```diff
@@ -2,132 +2,128 @@ import Foundation
...
-        let userClient = ONGUserClient.sharedInstance()
-        return isUserEnrolledForMobileAuth(userClient: userClient)
-    }
-    
-    func isUserEnrolledForMobileAuth(userClient: ONGUserClient) -> Bool {
-        if let userProfile = userClient.authenticatedUserProfile() {
-            return userClient.isUserEnrolled(forMobileAuth: userProfile)
+
+        // Prevent concurrent OTP mobile authentication flows at same time; https://onewelcome.atlassian.net/browse/SDKIOS-989
+        if (flowStarted) {
```
Comment (szymon-onewelcome) [inline ios/Classes/NativeBridge/Handlers/MobileAuthHandler.swift:29]: `guard !flowStarted else {`

File: ios/Classes/NativeBridge/Handlers/MobileAuthHandler.swift [hunk 4]
```diff
@@ -2,132 +2,128 @@ import Foundation
...
-    var confirmation: ((Bool) -> Void)?
-    var mobileAuthCompletion: ((Any?, SdkError?) -> Void)?
-        
-    fileprivate func handleConfirmationMobileAuth(_ cancelled: Bool) {
-        guard let confirmation = confirmation else { fatalError() }
-        
-        confirmation(cancelled)
-    }
+    var mobileAuthCallback: ((Bool) -> Void)?
+    var flowStarted: Bool = false
```
Comment (szymon-onewelcome) [inline ios/Classes/NativeBridge/Handlers/MobileAuthHandler.swift:14]: make `private`, also I would rename to sth like `isflowInProgress`

File: ios/Classes/NativeBridge/Handlers/MobileAuthHandler.swift [hunk 5]
```diff
@@ -2,132 +2,128 @@ import Foundation
...
-    var authenticatorType: MobileAuthAuthenticatorType?
-    var confirmation: ((Bool) -> Void)?
-    var mobileAuthCompletion: ((Any?, SdkError?) -> Void)?
-        
-    fileprivate func handleConfirmationMobileAuth(_ cancelled: Bool) {
-        guard let confirmation = confirmation else { fatalError() }
-        
-        confirmation(cancelled)
-    }
+    var mobileAuthCallback: ((Bool) -> Void)?
```
Comment (szymon-onewelcome) [inline ios/Classes/NativeBridge/Handlers/MobileAuthHandler.swift:13]: Make `private`

File: ios/Classes/NativeBridge/Handlers/MobileAuthHandler.swift [hunk 6]
```diff
@@ -2,132 +2,128 @@ import Foundation
...
+
+    func userClient(_ userClient: UserClient, didReceiveConfirmation confirmation: @escaping (Bool) -> Void, for request: MobileAuthRequest) {
+        BridgeConnector.shared?.toMobileAuthHandler.setMobileAuthCallback(didReceiveConfirmationChallenge: confirmation)
+        SwiftOneginiPlugin.flutterApi?.n2fOpenAuthOtp(message: request.message) {}
     }
 
-    func userClient(_: ONGUserClient, didReceive challenge: ONGPinChallenge, for request: ONGMobileAuthRequest) {
-       //@todo will need this for PUSH
+    func userClient(_ userClient: UserClient, didReceivePinChallenge challenge: PinChallenge, for request: MobileAuthRequest) {
+        //@todo will need this for PUSH
```
Comment (szymon-onewelcome) [inline ios/Classes/NativeBridge/Handlers/MobileAuthHandler.swift:100]: why not `// TODO:` ?

File: ios/Classes/NativeBridge/Handlers/MobileAuthHandler.swift [hunk 7]
```diff
@@ -2,132 +2,128 @@ import Foundation
...
     }
 
-    func userClient(_ userClient: ONGUserClient, didFailToHandle request: ONGMobileAuthRequest, authenticator: ONGAuthenticator?, error: Error) {
+    func userClient(_ userClient: UserClient, didFailToHandleRequest request: MobileAuthRequest, authenticator: Authenticator?, error: Error) {
+        BridgeConnector.shared?.toMobileAuthHandler.finishMobileAuthenticationFlow()
         SwiftOneginiPlugin.flutterApi?.n2fCloseAuthOtp {}
+
         if error.code == ONGGenericError.actionCancelled.rawValue {
-            mobileAuthCompletion?(false, SdkError(.authenticationCancelled))
+            self.handleMobileAuthCompletion(.failure(FlutterError(SdkError(.authenticationCancelled))))
```
Comment (szymon-onewelcome) [inline ios/Classes/NativeBridge/Handlers/MobileAuthHandler.swift:116]: I believe you don't need `self.` here

File: ios/Classes/NativeBridge/Handlers/MobileAuthHandler.swift [hunk 8]
```diff
@@ -2,132 +2,128 @@ import Foundation
...
         } else {
             let mappedError = ErrorMapper().mapError(error)
-            mobileAuthCompletion?(false, mappedError)
+            self.handleMobileAuthCompletion(.failure(FlutterError(mappedError)))
```
Comment (szymon-onewelcome) [inline ios/Classes/NativeBridge/Handlers/MobileAuthHandler.swift:119]: I believe you don't need `self.` here

File: ios/Classes/NativeBridge/Handlers/MobileAuthHandler.swift [hunk 9]
```diff
@@ -2,132 +2,128 @@ import Foundation
...
         }
     }
 
-    func userClient(_ userClient: ONGUserClient, didHandle request: ONGMobileAuthRequest, authenticator: ONGAuthenticator?, info customAuthenticatorInfo: ONGCustomInfo?) {
+    func userClient(_ userClient: UserClient, didHandleRequest request: MobileAuthRequest, authenticator: Authenticator?, info customAuthenticatorInfo: CustomInfo?) {
+        BridgeConnector.shared?.toMobileAuthHandler.finishMobileAuthenticationFlow()
         SwiftOneginiPlugin.flutterApi?.n2fCloseAuthOtp {}
-        mobileAuthCompletion?(request.message, nil)
+
+        self.handleMobileAuthCompletion(.success)
```
Comment (szymon-onewelcome) [inline ios/Classes/NativeBridge/Handlers/MobileAuthHandler.swift:127]: I believe you don't need `self.` here

**Review by szymon-onewelcome — CHANGES_REQUESTED** [review]

File: ios/Classes/NativeBridge/Handlers/MobileAuthHandler.swift (hunk 1 above)
Comment (Archifer) [reply ios/Classes/NativeBridge/Handlers/MobileAuthHandler.swift:9]: 1. the mobileAuthCallback is the one that we need to accept or deny during the mobileauthentication flow; while the other completions are the completions for the functions themselves 
2. bool is the one we receive from the native SDK 
3. replaced the `didReceiveConfirmationChallenge` with _

File: ios/Classes/NativeBridge/Handlers/MobileAuthHandler.swift (hunk 2 above)
Comment (Archifer) [reply ios/Classes/NativeBridge/Handlers/MobileAuthHandler.swift:23]: done

File: ios/Classes/NativeBridge/Handlers/MobileAuthHandler.swift (hunk 3 above)
Comment (Archifer) [reply ios/Classes/NativeBridge/Handlers/MobileAuthHandler.swift:29]: done

File: ios/Classes/NativeBridge/Handlers/MobileAuthHandler.swift (hunk 5 above)
Comment (Archifer) [reply ios/Classes/NativeBridge/Handlers/MobileAuthHandler.swift:13]: done

File: ios/Classes/NativeBridge/Handlers/MobileAuthHandler.swift (hunk 6 above)
Comment (Archifer) [reply ios/Classes/NativeBridge/Handlers/MobileAuthHandler.swift:100]: changed; it was already like this before the rework but you were right

File: ios/Classes/NativeBridge/Handlers/MobileAuthHandler.swift (hunk 9 above)
Comment (Archifer) [reply ios/Classes/NativeBridge/Handlers/MobileAuthHandler.swift:127]: done

File: ios/Classes/NativeBridge/Handlers/MobileAuthHandler.swift (hunk 8 above)
Comment (Archifer) [reply ios/Classes/NativeBridge/Handlers/MobileAuthHandler.swift:119]: done

**Pull request merged**